doc/api18970.md

```diff
@@ -0,0 +1,511 @@
+`k_sleep`         | yes        | yes   | no      | no     | no
+`k_sem_take`      | yes        | yes   | yes     | no     | no
+`k_thread_create` | ?          | ?     | ?       | ?      | yes
+
+### reschedule
+
+The reschedule attribute is used on a function that can reach a
+reschedule point within its execution.
+
+#### Commentary
+
+The significance of this attribute is that when a rescheduling function is
+invoked by a thread it is possible for that thread to be suspended as a
+consequence of a higher-priority thread being made ready.  Whether the
+suspension actually occurs depends on the operation associated with the
+reschedule point and the relative priorities of the invoking thread and
+the head of the ready queue.
+
+Note that in the case of timeslicing, or reschedule points executed from
+interrupts, any thread may be suspended in any function.
+
+Functions that are not **reschedule** may be invoked from either thread
+or interrupt context.
+
+Functions that are **reschedule** may be invoked from thread context.
+
+Functions that are **reschedule** but not **sleep** may be invoked from
+interrupt context.
+
+### sleep
+
+The sleep attribute is used on a function that can cause the invoking
+thread to sleep.
+
+Functions that are **sleep** are implicitly **reschedule**.
+
+Functions that are **sleep** may be invoked from thread context.
+
+Functions that are **sleep** may be invoked from interrupt context if
+and only if invoked in **no-wait** mode.
+
+#### Commentary
+
+This attribute is of relevance specifically when considering application
+that use only cooperative threads, because the kernel will not replace a
+running cooperative thread at a reschedule point unless that thread has
+explicitly invoked an operation that caused it to sleep.
+
+Note that this attribute does not imply the function will sleep
+unconditionally, but that the state of the system may require the
+invoking thread to suspend, wait, or invoke `k_yield()` before it can
+complete its operation.  This behavior may be mediated by **no-wait**.
+
+### no-wait
+
+The no-wait attribute is used on a function to indicate that a parameter
+to the function can force an execution path that will not cause the
+invoking thread to sleep.
+
+Functions with this attribute may be invoked from interrupt context when
+the parameter selects the no-wait path.
+
+Functions that are **no-wait** functions are implicitly **sleep**.
+
+### isr-ok
+
+The isr-ok attribute is used on a function to indicate that it can be
+called from interrupt context to the function can force an execution
+path that will not cause the invoking thread to sleep.
+
+This attribute is intended for **sleep** functions that may be
+indirectly invoked from interrupt context with arguments that would
+attempt to cause the invoking thread to sleep.
+
+Functions that are **isr-ok** may be invoked from interrupt context, and
+will return an error if they would have tried to sleep.
```

Comment:
Are you sure about making this a hard universal rule? Why not allow certain functions to conform to the historical Zephyr style of not checking for this type of error except via asserts, to avoid overhead?

Reply:
Because that historical style is exactly why we need this sort of thing.

Take `k_sem_take()`.  You can't call it from an interrupt unless you invoke it in no-wait mode.  If you do the kernel will panic--either because of an assert, or because it's wandered into the fog.  It's "OK" to not return an error because the developer will always know whether the call is in the body of an interrupt handler or a function invoked from a thread, and will have rigorously tested all execution paths with assertions enabled and will be justified in their confidence that the application behaves correctly.

Now consider `gpio_pin_write()`.  You should *always* be able to call that from thread and interrupt context, because it's safe to do that with most SOC controllers and making it illegal would be a major pain in driver code where you need to do something like deassert a signal.

But if the underlying controller identified in a sensor's devicetree node is accessed with I2C then when the driver deasserts the signal from an interrupt handler the GPIO implementation absolutely cannot try to take the semaphore that guarantees mutex of the bus.  So `gpio_pin_write()` must be `isr-ok`, and rather than bring the system down must instead return an error to indicate that it didn't do what it was asked to do.  And the code must propagate that error back to its caller, which must check it and complete its responsibilities some other way.

Now this does reveal an ambiguity that's part of why `k_sem_take()` is not `isr-ok`: because if it was, it would have to do a runtime check, which would get people upset (though the coding guidelines are going to require it anyway).  `gpio_pin_write()` is also not `sleep` or even `reschedule`, even though in certain controllers it will have to sleep.

That's why, once all this stuff gets in place and topic-gpio starts moving forward again, the GPIO API will have to be extended so each gpio controller can provide information about which of its operations are `sleep` so applications can take that into account when interacting with the devices.   (See #17827, which is the fundamental need I've had for the last six months that's motivating this whole effort, and #21090 where the transition operations must be marked as to whether they might sleep so the generic interface can safely avoid panics.)